### PR TITLE
Advertise TopologyManger policy and scope as Attributes in NRT api v1alpha2

### PR DESCRIPTION
**Pull request by PiotrProkop**
We want to advertise TopologyManager policy and scope as top level attributes in noderesourcetopology-api `v1alpha2`.
Based on this PR https://github.com/kubernetes-sigs/scheduler-plugins/pull/488.

### CODE REVIEW

**Comment by PiotrProkop** [comment]
I'll add unit tests for this tomorrow. @ffromani @swatisehgal please review.

**Review by ffromani — COMMENTED** [review]
thanks for the quick PR! looks straightforward, suggest possible improvements inside.

File: pkg/topologypolicy/topology-policy.go [hunk 1]
```diff
@@ -63,3 +70,16 @@ func detectPolicyContainerScope(policy string) v1alpha2.TopologyManagerPolicy {
 		return v1alpha2.None
 	}
 }
+
+func DetectTopologyAttributes(policy string, scope string) v1alpha2.AttributeList {
```
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply pkg/topologypolicy/topology-policy.go:74]: will do.

File: cmd/nfd-topology-updater/main.go [hunk 2]
```diff
@@ -88,11 +87,13 @@ func main() {
 		klog.Exitf("unsupported URI scheme: %v", u.Scheme)
 	}
 
-	tmPolicy := string(topologypolicy.DetectTopologyPolicy(klConfig.TopologyManagerPolicy, klConfig.TopologyManagerScope))
+	tmPolicy := string(klConfig.TopologyManagerPolicy)
 	klog.Infof("detected kubelet Topology Manager policy %q", tmPolicy)
+	tmScope := string(klConfig.TopologyManagerScope)
+	klog.Infof("detected kubelet Topology Manager scope %q", tmScope)
```
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply cmd/nfd-topology-updater/main.go:93]: Makes sense, changing it now.

File: pkg/topologypolicy/topology-policy.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ffromani) [reply pkg/topologypolicy/topology-policy.go:74]: Thanks! if you move, please also rename this function! The code is just fine, but it is not detecting, rather constructing/creating

File: cmd/nfd-topology-updater/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply cmd/nfd-topology-updater/main.go]: I have put logging policy, scope and nodeName to new func called `newStaticNodeInfo`.

File: pkg/topologypolicy/topology-policy.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply pkg/topologypolicy/topology-policy.go]: moved and changed, and got rid of `AttributeName` type alias which was unused :)

File: pkg/nfd-topology-updater/nfd-topology-updater.go [hunk 3]
```diff
@@ -60,6 +68,17 @@ type NfdTopologyUpdater interface {
 type staticNodeInfo struct {
 	nodeName string
 	tmPolicy string
+	tmScope  string
+}
+
+func newStaticNodeInfo(policy, scope string) *staticNodeInfo {
```
Comment (ffromani) [inline pkg/nfd-topology-updater/nfd-topology-updater.go:74]: nit/nonblocking: I'd just return a struct (not a pointer to a struct) and trust the compiler.
In general, returning pointers *may* allow faster copying of objects (but it should actually be measured) but also contributes to the GC pressure (which, OTOH, should also be measured).
What I'm trying to say is that returning pointers is not necessarily the faster option,

**Review by ffromani — COMMENTED** [review]
looks fine. I'd like to see some form of test coverage (proably integration/e2e) before LGTM

File: pkg/nfd-topology-updater/nfd-topology-updater.go (hunk 3 above)
Comment (ffromani) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:74]: note: I'd NOT resubmit just for this.

Comment (PiotrProkop) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:74]: I want to add unit tests, so I'll change it when I submit them.

**Comment by PiotrProkop** [comment]
@ffromani I had to change the way how we load Kubelet config when read from file to make e2e test working, previously we were just reading config from file and unamarshalling it to struct. This approach didn't fill out default values but were skipping it based on `omitempty` tag. In e2e tests we are reading Kubelet config via http, which is serving Kubeletconfig with all values filled 😄

File: test/e2e/utils/noderesourcetopology.go [hunk 4]
```diff
@@ -259,3 +279,12 @@ func isValidResourceList(zoneName string, resources v1alpha2.ResourceInfoList) b
 	}
 	return foundCpu
 }
+
+func containsAttribute(attributes v1alpha2.AttributeList, attribute v1alpha2.AttributeInfo) bool {
+	for _, attr := range attributes {
+		if reflect.DeepEqual(attr, attribute) {
```
Comment (ffromani) [inline test/e2e/utils/noderesourcetopology.go:285]: minor: reflect is no bother, but can't we just use '=='?

File: pkg/utils/kubeconf/kubelet_config_file.go [hunk 5]
```diff
@@ -17,23 +17,40 @@ limitations under the License.
...
 
 	kubeletconfigv1beta1 "k8s.io/kubelet/config/v1beta1"
+	kubeletconfigscheme "k8s.io/kubernetes/pkg/kubelet/apis/config/scheme"
+	"k8s.io/kubernetes/pkg/kubelet/kubeletconfig/configfiles"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
 )
 
 // GetKubeletConfigFromLocalFile returns KubeletConfiguration loaded from the node local config
+// based on https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L337
+// it fills empty fields with default values
```
Comment (ffromani) [inline pkg/utils/kubeconf/kubelet_config_file.go:30]: this looks indeed more correct, thanks for improving this function

**Review by ffromani — COMMENTED** [review]
/lgtm
very nice work @PiotrProkop !

File: test/e2e/utils/noderesourcetopology.go (hunk 4 above)
Comment (PiotrProkop) [reply test/e2e/utils/noderesourcetopology.go:285]: I was thinking about doing:
```go
if attr.Name == attribute.Name && attr.Value == attribute.Value {
    return true
}
```
but settle on shorter one.

File: pkg/utils/kubeconf/kubelet_config_file.go (hunk 5 above)
Comment (PiotrProkop) [reply pkg/utils/kubeconf/kubelet_config_file.go:30]: 🎸

**Comment by marquiz** [comment]
Thanks @PiotrProkop for the PR. I'll check this next week

File: test/e2e/utils/noderesourcetopology.go (hunk 4 above)
Comment (ffromani) [reply test/e2e/utils/noderesourcetopology.go:285]: I would have used the explicit == version, but I don't have strong feelings so works for me.

Comment (marquiz) [reply test/e2e/utils/noderesourcetopology.go:285]: My thinking also that `attr == attribute` should've done the job but especially in the e2e tests I don't have that strong feelings either so let's go with this...

File: pkg/utils/kubeconf/kubelet_config_file.go [hunk 6]
```diff
@@ -17,23 +17,40 @@ limitations under the License.
...
 func GetKubeletConfigFromLocalFile(kubeletConfigPath string) (*kubeletconfigv1beta1.KubeletConfiguration, error) {
-	kubeletBytes, err := os.ReadFile(kubeletConfigPath)
+	const errFmt = "failed to load Kubelet config file %s, error %v"
```
Comment (marquiz) [inline pkg/utils/kubeconf/kubelet_config_file.go:32]: nit: wrap errors with `%w`

```suggestion
	const errFmt = "failed to load Kubelet config file %s: %w"
```

Comment (PiotrProkop) [reply pkg/utils/kubeconf/kubelet_config_file.go]: fixed

**Review by marquiz — APPROVED** [review]
Thanks @PiotrProkop, looks good to me now

**Comment by PiotrProkop** [comment]
@ffromani can you LGTM once more? I had to fix a small nit.

**Pull request merged**